core/shared/src/main/scala/scalaz/zio/Exit.scala

```diff
@@ -206,16 +226,28 @@ object Exit extends Serializable {
 
   object Cause extends Serializable {
 
-    final def fail[E](error: E): Cause[E] = Fail(error)
-
-    final def die(defect: Throwable): Cause[Nothing] = Die(defect)
-
-    final val interrupt: Cause[Nothing] = Interrupt
+    final def fail[E](error: E): Cause[E]                            = Fail(error)
+    final def die(defect: Throwable): Cause[Nothing]                 = Die(defect)
+    final val interrupt: Cause[Nothing]                              = Interrupt
+    final def finalizerErrors(cause: Cause[Nothing]): Cause[Nothing] = FinalizerErrors(cause)
```

Comment:
Isn't an exception in the finalizer a reason to die and then this could rather be encoded as `FinalizerError <: Throwable`, with only a small change in `unwindStack` necessary?

Reply:
This refers only to errors that occured during `unwindStack` - when defect in release caused the exit it goes into `Die` as usual.
Unfortunately, we have corner cases that allow a any `Cause` to arise from finalizer in unwindStack, not just Throwable from defects. Say:
`io.ensuring(IO.interrupt)`, `io.ensuring(Fiber.joinAll(...))` or just `io.ensuring(IO.halt(anyCause))`

Reply:
Well, we could wrap everything from `unwindStack` into a special Throwable, then typecase in Cause to distinguish real and unwind errors, this would break much less, which might be preferrable, if a bit dirty